File: LeetCode/1464. Maximum Product of Two Elements in an Array/solution.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        firstMax = {"value": -inf, "index": None}
        secondMax = {"value": -inf, "index": None}

        for i, num in enumerate(nums):
            if num > firstMax["value"]:
                secondMax["value"] = firstMax["value"]
                secondMax["index"] = firstMax["index"]
                firstMax["value"] = num
                firstMax["index"] = i
            elif num > secondMax["value"]:
                secondMax["value"] = num
                secondMax["index"] = i

            # Debug ------------------------
            # print(num, firstMax, secondMax)

        return (firstMax["value"] - 1) * (secondMax["value"] - 1)

File: LeetCode/1464. Maximum Product of Two Elements in an Array/Solution.java
import java.util.HashMap;

class Solution {
    public int maxProduct(int[] nums) {
        HashMap<String, Integer> firstMax = new HashMap<String, Integer>();
        firstMax.put("value", -999999999);
        firstMax.put("index", -1);

        HashMap<String, Integer> secondMax = new HashMap<String, Integer>();
        secondMax.put("value", -999999999);
        secondMax.put("index", -1);

        for (int i = 0; i < nums.length; i++) {

            if (nums[i] > firstMax.get("value")) {
                secondMax.put("value", firstMax.get("value"));
                secondMax.put("index", firstMax.get("index"));
                
                firstMax.put("value", nums[i]);
                firstMax.put("index", i);

            } else if (nums[i] > secondMax.get("value")) {
                secondMax.put("value", nums[i]);
                secondMax.put("index", i);
            }

        }
        
        return (firstMax.get("value") - 1) * (secondMax.get("value") - 1);
    }
}
